include/openssl/hpke.h

```diff
@@ -0,0 +1,466 @@
+ * @param label is a buffer to provide separation between secrets
+ * @param labellen is the length of the above
+ * @return 1 for good, 0 for error
+ *
+ * The context has to have been used already for one encryption
+ * or decryption for this to work (as this is based on the negotiated
+ * "exporter_secret" estabilshed via the HPKE operation).
+ */
+int OSSL_HPKE_CTX_export(OSSL_HPKE_CTX *ctx,
+                         unsigned char *secret,
+                         size_t secret_len,
+                         const unsigned char *label,
+                         size_t labellen);
+
+/**
+ * @brief generate a key pair
+ * @param libctx is the context to use (normally NULL)
+ * @param propq is a properties string
+ * @param mode is the mode (currently unused)
+ * @param suite is the ciphersuite (currently unused)
+ * @param ikm is IKM, if supplied
+ * @param ikmlen is the length of IKM, if supplied
+ * @param pub is the public value
+ * @param publen is the size of the public key buffer (exact length on output)
+ * @param priv is the private key
+ * @return 1 for success, other for error (error returns can be non-zero)
+ *
+ * Used for entities that will later receive HPKE values to
+ * decrypt or that want a private key for an AUTH mode. Currently,
+ * only the KEM from the suite is significant here.
+ * The ``pub`` output will typically be published so that
+ * others can encrypt to the private key holder using HPKE.
+ * (Or authenticate HPKE values from that sender.)
+ */
+int OSSL_HPKE_keygen(OSSL_LIB_CTX *libctx, const char *propq,
+                     unsigned int mode, OSSL_HPKE_SUITE suite,
+                     const unsigned char *ikm, size_t ikmlen,
+                     unsigned char *pub, size_t *publen, EVP_PKEY **priv);
+
+/**
+ * @brief check if a suite is supported locally
+ * @param suite is the suite to check
+ * @return 1 for success, other for error (error returns can be non-zero)
+ */
+int OSSL_HPKE_suite_check(OSSL_HPKE_SUITE suite);
+
+/**
+ * @brief get a (possibly) random suite, public key and ciphertext for GREASErs
+ * @param libctx is the context to use (normally NULL)
+ * @param propq is a properties string
+ * @param suite_in specifies the preferred suite or NULL for a random choice
+ * @param suite is the chosen or random suite
+ * @param pub a random value of the appropriate length for a sender public value
+ * @param pub_len is the length of pub (buffer size on input)
+ * @param cipher is a random value of the appropriate length for a ciphertext
+ * @param cipher_len is the length of cipher
+ * @return 1 for success, otherwise failure
+ *
+ * If suite_in is provided that will be used (if supported). If
+ * suite_in is NULL, a random suite (from those supported) will
+ * be selected. In all cases the output pub and cipher values
+ * will be appropriate random values for the selected suite.
+ */
+int OSSL_HPKE_good4grease(OSSL_LIB_CTX *libctx, const char *propq,
```

Comment:
Not sure about the name 'good4'. Make the function describe what it does? Is it returning random values..
Why is pub_len a pointer  but cipher_len is not?
Please use ciphertext or ct as the name, cipher is a bit confusing.

Reply:
will do

Reply:
better name suggestion accepted below